analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -287,20 +287,25 @@ operation.Kind switch
 
         private static ProgramState LearnBranchingConstraints(ControlFlowBranch branch, ProgramState state, IOperation branchValue)
         {
-            if (branchValue.Kind == OperationKindEx.Conversion)
-            {
-                var operand = IConversionOperationWrapper.FromOperation(branchValue).Operand;
-                state = LearnBranchingConstraints(branch, state, operand);
-                return state.SetOperationValue(branchValue, state[operand]);    // Propagate value from operand to conversion operation itself
-            }
-            else
-            {
-                var constraint = BoolConstraint.From((branch.Source.ConditionKind == ControlFlowConditionKind.WhenTrue) == branch.IsConditionalSuccessor);
-                state = state.SetOperationConstraint(branchValue, constraint);
-                return branchValue.TrackedSymbol() is { } symbol
-                    ? state.SetSymbolConstraint(symbol, constraint)
-                    : state;
-            }
+            var trueBranch = (branch.Source.ConditionKind == ControlFlowConditionKind.WhenTrue) == branch.IsConditionalSuccessor;
+            state = state.SetOperationConstraint(branchValue, BoolConstraint.From(trueBranch));
+            return LearnBranchingConstraintsFromOperation(state, branchValue, !trueBranch);
+        }
+
+        private static ProgramState LearnBranchingConstraintsFromOperation(ProgramState state, IOperation operation, bool useOpposite)
+        {
+            return operation.TrackedSymbol() is { } symbol
+                ? state.SetSymbolConstraint(symbol, BoolConstraint.From(!useOpposite))
+                : operation.Kind switch
+                {
+                    OperationKindEx.Binary => Binary.LearnBranchingConstraint(state, As(IBinaryOperationWrapper.FromOperation), useOpposite),
+                    OperationKindEx.Conversion => LearnBranchingConstraintsFromOperation(state, As(IConversionOperationWrapper.FromOperation).Operand, useOpposite),
+                    OperationKindEx.Unary when operation.ToUnary() is { OperatorKind: UnaryOperatorKind.Not } unary => LearnBranchingConstraintsFromOperation(state, unary.Operand, !useOpposite),
+                    _ => state
```

Comment:
That is not a learning situation, because there's nothing to learn.
`true` literal has `True` constraint
`a` is assigned `true` therefore a has assigned `True` constraint
`if` condition is evaluated as "always true"

It should already be supported.